JavaIntegration/src/main/java/org/openzen/zencode/java/module/converters/JavaNativeTypeConverter.java

```diff
@@ -76,104 +85,152 @@ else if (annotatedType.isAnnotationPresent(ZenCodeType.NullableUSize.class))
-		if (type instanceof Class) {
-			Class<?> classType = (Class<?>) type;
-			if (unsigned) {
-				if (unsignedByClass.containsKey(classType)) {
-					return unsignedByClass.get(classType);
-				} else {
-					throw new IllegalArgumentException("This class cannot be used as unsigned: " + classType);
-				}
-			} else if (classType.isArray()) {
-				return typeConversionContext.registry.getArray(loadType(context, classType.getComponentType(), false, false), 1);
-			} else if (classType.isAnnotationPresent(FunctionalInterface.class)) {
-				return loadFunctionalInterface(context, classType, new AnnotatedElement[0]);
-			}
+	private TypeID loadType(TypeVariableContext context, JavaAnnotatedType type, boolean unsigned) {
 
-			if (typeByClass.containsKey(classType))
-				return typeByClass.get(classType);
+		final JavaAnnotatedType.ElementType elementType = type.getElementType();
 
-			HighLevelDefinition definition = javaNativeModule.addClass(classType);
-			final List<TypeID> s = new ArrayList<>();
-			for (TypeVariable<? extends Class<?>> typeParameter : classType.getTypeParameters()) {
-				s.add(typeConversionContext.registry.getGeneric(context.get(typeParameter)));
+		try {
+			switch (elementType) {
+				case ANNOTATED_PARAMETERIZED_TYPE:
+					return loadAnnotatedParameterizedType(context, (AnnotatedParameterizedType) type.getAnnotatedElement(), unsigned);
+				case ANNOTATED_TYPE:
+					return loadAnnotatedType(context, (AnnotatedType) type.getAnnotatedElement(), unsigned);
+				case CLASS:
+					return loadClass(context, (Class<?>) type.getType(), unsigned);
+				case GENERIC_ARRAY:
+					return loadGenericArray(context, (GenericArrayType) type.getType(), unsigned);
+				case PARAMETERIZED_TYPE:
+					return loadParameterizedType(context, (ParameterizedType) type.getType());
+				case TYPE_VARIABLE:
+					return loadTypeVariable(context, (TypeVariable<?>) type.getType());
+				case WILDCARD:
+					return loadWildcard();
 			}
 
-			return typeConversionContext.registry.getForDefinition(definition, s.toArray(TypeID.NONE));
-		} else if (type instanceof ParameterizedType) {
-			ParameterizedType parameterizedType = (ParameterizedType) type;
-			Class<?> rawType = (Class<?>) parameterizedType.getRawType();
-			if (rawType.isAnnotationPresent(FunctionalInterface.class))
-				return loadFunctionalInterface(context, rawType, (AnnotatedElement[]) parameterizedType.getActualTypeArguments());
+			throw new IllegalArgumentException("Invalid type " + elementType + ": not yet implemented or foolery");
+		} catch (final IllegalArgumentException e) {
+			throw new IllegalArgumentException("Unable to analyze type: " + type, e);
+		}
+	}
+
+	private TypeID loadAnnotatedParameterizedType(TypeVariableContext context, AnnotatedParameterizedType type, boolean unsigned) {
+		final ParameterizedType parameterizedType = this.getTypeIfValid(JavaAnnotatedType.of(type.getType()), JavaAnnotatedType.ElementType.PARAMETERIZED_TYPE);
```

Comment:
The contract of `getType` in `AnnotatedParameterizedType` makes no assumption on the return value of the method, so I figured that we might as well verify that the preconditions we require hold, just in case